### DISCUSSION TITLE
Python: better experience for plain dictionaries (PEP 655)

### DISCUSSION THREAD

**Discussion by antdking** [pulumi/pulumi#11500] [Ideas]
Firstly, thanks for making Pulumi, I'm using (and abusing) this day-to-day, and it's shaping up to be a great product.

pleasantries aside, onto the woe

The experience using Pulumi with Python isn't fantastic when compared to Typescript, mainly due to needing to use special classes to get type hints.
This results in verbose definitions, more boilerplate than the data we care about.

<details>
<summary>Let's use k8s' Ingress as an example:</summary>

```py
import pulumi_kubernetes as k8s

ingress = k8s.networking.v1.Ingress(
        SERVICE,
        metadata=k8s.meta.v1.ObjectMetaArgs(
            annotations={
              ...
            },
            namespace=namespace,
        ),
        spec=k8s.networking.v1.IngressSpecArgs(
            tls=[
                k8s.networking.v1.IngressTLSArgs(hosts=[PUBLIC_DOMAIN], secret_name="ingress-be-cert-public"),
            ],
            rules=[
                k8s.networking.v1.IngressRuleArgs(
                    host=PUBLIC_DOMAIN,
                    http=k8s.networking.v1.HTTPIngressRuleValueArgs(
                        paths=[
                            k8s.networking.v1.HTTPIngressPathArgs(
                                path="/",
                                path_type="Prefix",
                                backend=k8s.networking.v1.IngressBackendArgs(
                                    service=k8s.networking.v1.IngressServiceBackendArgs(
                                        name=SERVICE,
                                        port=k8s.networking.v1.ServiceBackendPortArgs(
                                            name="http",
                                        ),
                                    )
                                ),
                            )
                        ]
                    ),
                )
            ],
        ),
    )
```

</details>

The Ingress defines a TLS certificate to use/generate, and a basic HTTP rule to forward all traffic from `PUBLIC_DOMAIN` to a service `SERVICE`.
Unfortunately, **it's hard to see this with lots of nested classes** (though can be mitigated slightly by aliasing the data objects).

<details>
<summary>Compare this to using dictionary literals (which are supported at runtime already):</summary>

```py
ingress = k8s.networking.v1.Ingress(
        SERVICE,
        metadata={
            "annotations": {
              ...
            },
            "namespace": namespace,
        },
        spec={
            "tls": [
                {"hosts": [PUBLIC_DOMAIN], "secretName": "ingress-be-cert-public"},
            ],
            "rules": [
                {
                    "host": PUBLIC_DOMAIN,
                    "http": {
                        "paths": [
                            {
                                "backend": {
                                    "service": {
                                        "name": SERVICE,
                                        "port": {
                                            "name": "http",
                                        },
                                    },
                                },
                                "path": "/",
                                "pathType": "Prefix",
                            }
                        ]
                    },
                }
            ],
        },
    )
```
</details>

<details>
<summary>or `dict(...)`:</summary>

```py
Ingress(
        SERVICE,
        metadata=dict(
            annotations={
                ...
            },
            namespace=namespace,
        ),
        spec=dict(
            tls=[
                dict(hosts=[PUBLIC_DOMAIN], secret_name="ingress-be-cert-public"),
            ],
            rules=[
                dict(
                    host=PUBLIC_DOMAIN,
                    http=dict(
                        paths=[
                            dict(
                                path="/",
                                path_type="Prefix",
                                backend=dict(
                                    service=dict(
                                        name=SERVICE,
                                        port=dict(
                                            name="http",
                                        ),
                                    )
                                ),
                            )
                        ]
                    ),
                )
            ],
        ),
    )
```
</details>

We end up with less cruft in our definitions.

I did a little work last night to integrate [PEP 655](https://peps.python.org/pep-0655/) based TypedDict type hints, and it went surprisingly well, with hints + type errors throwing in vscode (via pylance) as expected.

I think it'd be good to look into this further, so we can begin writing cleaner code, akin to how Pulumi + Typescript works.

[Somewhat hacky implementation](https://github.com/pulumi/pulumi/compare/master...antdking:pulumi:feat/codegen/python-typed-dict).
The implementation relies on `typing_extensions`, and should be compatible back to python 3.7, with mypy and pylance.

Do you think there'd be appetite for picking this functionality up?

_I'm not overly familiar with the codebase, and haven't been able to add additional tests. As it is, I'm happy to have learnt more on how `codegen` works, and it's my first time with GoLang._

**Comment by Frassle** [comment]
cc @kpitzen 

We had discussed this internally recently, we suspect that PEP 655 is the way to go for good UX for Pulumi programs in Python, so yes there is definitely appetite for picking this up.

**Comment by julienp** [comment]
We've shipped TypedDict based input types for AWS and Kubernetes SDKs, with other SDKs following in the coming days and weeks. See also https://github.com/pulumi/pulumi/issues/12689

Read the blog about the changes https://www.pulumi.com/blog/pulumi-loves-python/
